contributors/devel/sig-architecture/api_changes.md

```diff
@@ -154,18 +159,42 @@ without changing the API version, so you can simply change it to:
 // Still API v6.
 type Frobber struct {
   Height int    `json:"height"`
-  Width  int    `json:"width"`
+  Width  *int   `json:"width"`  // optional, defaults to 42
   Param  string `json:"param"`
 }
 ```
 
+Note that fields which are added MUST be optional.  Optional fields must be
+nil-able (pointers or slices or (rarely) maps).
+
 The onus is on you to define a sane default value for `Width` such that rules
 #1 and #2 above are true - API calls and stored objects that used to work must
 continue to work.
 
+For newly added fields, the default behavior must be defined.  If this is a net
+new capability which did not exist before (i.e. a user must explicitly
+set it to change behavior), the default value should be "nothing" (`nil` in
+Go), meaning the feature is not activated. If this new field is extending an
```

Comment:
> "nothing"

When we expose configuration through the API, the existing behavior often not "nothing". Examples are InitialDelaySeconds for probes or AllowPrivilegeEscalation. By not exposing the value of new default fields, users have to have knowledge of what the existing behavior was to figure out what's going on. This also generally hurts debuggability.

Additionally the statement proposes that we can't turn non-breaking features on by default which doesn't sound very satisfying.

Reply:
I agree that it's a usability problem to require the default to always be "" or something.

Reply:
Devil's advocate: Users who know about the field can look at the API docs.  Users who don't know, don't care.  In this sense it is more compatible because people ALREADY have to figure out the defaults if they hit an API server that doesn't know that field.  Users have to handle nil anyway.  This makes old servers and new servers equally awkward.